imap_processing/hit/l0/decom_hit.py

```diff
@@ -0,0 +1,464 @@
+    These are assembled from the binary science_data field in the
+    xarray dataset, which is a 1D array of science data from all
+    packets in the file, using packet sequence grouping flags.
+
+        First packet has a sequence flag = 1
+        Next 18 packets have a sequence flag = 0
+        Last packet has a sequence flag = 2
+
+    The science frame is further categorized into
+    L1A data products -> count rates and event data.
+
+        The first six packets contain count rates data
+        The last 14 packets contain pulse height event data
+
+    These groups are added to the dataset as count_rates_binary
+    and pha_binary.
+
+    Parameters
+    ----------
+    sci_dataset : xr.Dataset
+        Xarray Dataset for science data (APID 1252).
+
+    Returns
+    -------
+    sci_dataset : xr.Dataset
+        Updated xarray dataset with binary count rates and pulse
+        height event data per valid science frame added as new
+        data variables.
+    """
+    # TODO: Figure out how to handle partial science frames at the
+    #  beginning and end of CCSDS files. These science frames are split
+    #  across CCSDS files and still need to be processed. Only discard
+    #  incomplete science frames in the middle of the CCSDS file.
+    #  The code currently skips all incomplete science frames.
+
+    # Initialize lists to store data from valid science frames
+    count_rates_binary = []
+    pha_binary = []
+    epoch_science_frame = []
+
+    science_frame_start = 0
+    while science_frame_start + 20 <= len(sci_dataset.epoch):
+        # Extract chunks for the current science frame (20 packets)
+        seq_flgs_chunk = sci_dataset.seq_flgs[
+            science_frame_start : science_frame_start + 20
+        ].values
+        src_seq_ctr_chunk = sci_dataset.src_seq_ctr[
+            science_frame_start : science_frame_start + 20
+        ].values
+        science_data_chunk = sci_dataset.science_data[
+            science_frame_start : science_frame_start + 20
+        ]
+        epoch_data_chunk = sci_dataset.epoch[
+            science_frame_start : science_frame_start + 20
+        ]
+
+        if is_valid_science_frame(seq_flgs_chunk, src_seq_ctr_chunk):
+            # Append valid data to lists
+            count_rates_binary.append("".join(science_data_chunk.data[0:6]))
+            pha_binary.append("".join(science_data_chunk.data[6:]))
+            # Just take first packet's epoch for the science frame
+            epoch_science_frame.append(epoch_data_chunk[0])
+            science_frame_start += 20  # Move to the next science frame
+        else:
+            print(
+                f"Invalid science frame found with starting packet index = "
+                f"{science_frame_start}"
+            )
+            # Skip science frame and move on to the next science frame packets
+            # Get index for the first packet in the next science frame
+            start = get_starting_packet_index(
```

Comment:
you are kind of doing what I mention above but instead of looking for the next `1`, you can find all of 1s ahead and then check if it's valid science frame.